### ISSUE TITLE
Panic when joining room

### ISSUE THREAD

**Issue by HenrikSolver** [matrix-org/dendrite#1401]
Running  a build from b9caccbce89ae1739218b2934fca7f5d38f3fe46 and joining #ping.maunium.net gave this.

```
        Request panicked!
goroutine 1646932 [running]:
runtime/debug.Stack(0xc0003e30a0, 0xc0018f3230, 0xc0003e33b0)
        runtime/debug/stack.go:24 +0x9d
github.com/matrix-org/util.Protect.func1.1(0xc001b56800, 0x1282e20, 0xc0031100e0)
        github.com/matrix-org/util@v0.0.0-20200807132607-55161520e1d4/json.go:98 +0x15d
panic(0x101bd80, 0xc003309700)
        runtime/panic.go:679 +0x1b2
github.com/matrix-org/gomatrixserverlib.RespState.Check(0x0, 0x0, 0xc00567c000, 0x235, 0x400, 0xc002192000, 0x335, 0x400, 0x1287a20, 0xc001f23d70, ...)
        github.com/matrix-org/gomatrixserverlib@v0.0.0-20200902135805-f7a5b5e89750/federationtypes.go:447 +0x1417
github.com/matrix-org/gomatrixserverlib.RespSendJoin.Check(0x1b56fb5, 0x1, 0xc00567c000, 0x236, 0x400, 0xc002192000, 0x335, 0x400, 0x0, 0x0, ...)
        github.com/matrix-org/gomatrixserverlib@v0.0.0-20200902135805-f7a5b5e89750/federationtypes.go:553 +0x197
github.com/matrix-org/dendrite/federationsender/internal/perform.joinContext.CheckSendJoinResponse(0xc000234cb0, 0xc000233500, 0x1287a20, 0xc001f23d70, 0x0, 0xc001198f00, 0x4a3, 0x4a3, 0xfced80, 0xc001d17b20, ...)
        github.com/matrix-org/dendrite@/federationsender/internal/perform/join.go:100 +0x1d4
github.com/matrix-org/dendrite/federationsender/internal.(*FederationSenderInternalAPI).performJoinUsingServer(0xc00019de50, 0x1287a20, 0xc001f23d70, 0xc0012e3c00, 0x1f, 0xc000cff6c0, 0x11, 0xc001f23e90, 0xc000cff6e6, 0xb, ...)
        github.com/matrix-org/dendrite@/federationsender/internal/perform.go:188 +0xa39
github.com/matrix-org/dendrite/federationsender/internal.(*FederationSenderInternalAPI).PerformJoin(0xc00019de50, 0x1287a20, 0xc001f23d70, 0xc001515500, 0xc001ec63d0)
        github.com/matrix-org/dendrite@/federationsender/internal/perform.go:70 +0x3e2
github.com/matrix-org/dendrite/roomserver/internal/perform.(*Joiner).performFederatedJoinRoomByID(0xc0002b5900, 0x1287a20, 0xc001f23d70, 0xc00309d400, 0xc0001d6008, 0xc0018f4d50)
        github.com/matrix-org/dendrite@/roomserver/internal/perform/perform_join.go:305 +0x12e
github.com/matrix-org/dendrite/roomserver/internal/perform.(*Joiner).performJoinRoomByID(0xc0002b5900, 0x1287a20, 0xc001f23d70, 0xc00309d400, 0xc001514900, 0x1, 0x4, 0x1)
        github.com/matrix-org/dendrite@/roomserver/internal/perform/perform_join.go:279 +0xcad
github.com/matrix-org/dendrite/roomserver/internal/perform.(*Joiner).performJoinRoomByAlias(0xc0002b5900, 0x1287a20, 0xc001f23d70, 0xc00309d400, 0x6, 0xc000cff6c8, 0x9, 0x0)
        github.com/matrix-org/dendrite@/roomserver/internal/perform/perform_join.go:137 +0x4b7
github.com/matrix-org/dendrite/roomserver/internal/perform.(*Joiner).performJoin(0xc0002b5900, 0x1287a20, 0xc001f23d70, 0xc00309d400, 0xc000d30d80, 0x1500000001287a20, 0xc0018f5000, 0x15752976cdc3e6ba)
        github.com/matrix-org/dendrite@/roomserver/internal/perform/perform_join.go:85 +0x2d5
github.com/matrix-org/dendrite/roomserver/internal/perform.(*Joiner).PerformJoin(0xc0002b5900, 0x1287a20, 0xc001f23d70, 0xc00309d400, 0xc0021bcce0)
        github.com/matrix-org/dendrite@/roomserver/internal/perform/perform_join.go:50 +0x53
github.com/matrix-org/dendrite/clientapi/routing.JoinRoomByIDOrAlias(0xc001b56a00, 0xc00344d040, 0x12a4840, 0xc00023b1e0, 0x12a0200, 0xc0001aaa20, 0xc000cff6e0, 0x11, 0xc0018f5501, 0xc001f23d70, ...)
        github.com/matrix-org/dendrite@/clientapi/routing/joinroom.go:77 +0x444
github.com/matrix-org/dendrite/clientapi/routing.Setup.func3(0xc001b56a00, 0xc00344d040, 0xef12e0, 0x1253dd0, 0x10a19c0, 0xc0003e3260)
        github.com/matrix-org/dendrite@/clientapi/routing/routing.go:103 +0x190
github.com/matrix-org/dendrite/internal/httputil.MakeAuthAPI.func1(0xc001b56900, 0x0, 0x0, 0x0, 0x0)
        github.com/matrix-org/dendrite@/internal/httputil/httpapi.go:63 +0x2c0
github.com/matrix-org/util.(*jsonRequestHandlerWrapper).OnIncomingRequest(0xc0000114a8, 0xc001b56900, 0xc0018f5760, 0xc000758000, 0x7fe9fab446d0, 0x0)
        github.com/matrix-org/util@v0.0.0-20200807132607-55161520e1d4/json.go:79 +0x33
github.com/matrix-org/util.MakeJSONAPI.func1(0x1282e20, 0xc0031100e0, 0xc001b56900)
        github.com/matrix-org/util@v0.0.0-20200807132607-55161520e1d4/json.go:141 +0x97
github.com/matrix-org/util.Protect.func1(0x1282e20, 0xc0031100e0, 0xc001b56800)
        github.com/matrix-org/util@v0.0.0-20200807132607-55161520e1d4/json.go:103 +0xa1
net/http.HandlerFunc.ServeHTTP(...)
        net/http/server.go:2007
github.com/matrix-org/dendrite/internal/httputil.MakeExternalAPI.func1(0x1282e20, 0xc0031100e0, 0xc001b56600)
        github.com/matrix-org/dendrite@/internal/httputil/httpapi.go:123 +0x20a
net/http.HandlerFunc.ServeHTTP(0xc00031f500, 0x1282e20, 0xc0031100e0, 0xc001b56600)
        net/http/server.go:2007 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0001f4480, 0x1282e20, 0xc0031100e0, 0xc001b56400)
        github.com/gorilla/mux@v1.8.0/mux.go:210 +0xe2
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0001f5740, 0x1282e20, 0xc0031100e0, 0xc001b56200)
        github.com/gorilla/mux@v1.8.0/mux.go:210 +0xe2
net/http.serverHandler.ServeHTTP(0xc00033c1c0, 0x1282e20, 0xc0031100e0, 0xc001b56200)
        net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc001db37c0, 0x1287960, 0xc00267bf40)
        net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
        net/http/server.go:2928 +0x384  context=missing panic="runtime error: index out of range [565] with length 565"
```

**Comment by neilalexander** [comment]
The panic will be fixed in gomatrixserverlib, but joining that room still won't work because of the state event `$wIf_9uvJqQIohAqk29FgMvNAb5GopaK_uVRyXiiE8SI`.

The event content is:
```
{
  "auth_events": [
    "$drGjohWCjpiD2DW-RrY6CZHzYV6jmihxffBaD8StYXM",
    "$Dh8ApAATX--ePjrk8_l2RraO0NtIBSAMnK8R20HSqlQ",
    "$xJXz-TMx6SrSfvSL5Hejo_Em72IRa4HAPNuVtOXrohE",
    "$qv1lfutiPmcEOSbDixX1nhJqevfWhQKdZpf0fq-3eZk"
  ],
  "content": {
    "membership": "invite"
  },
  "depth": 15869,
  "hashes": {
    "sha256": "PK1Ua1LjEFMPHS2bRz6Oce2CGK3nz4rzQzoHvsAZOVs"
  },
  "origin": "zemos.net",
  "origin_server_ts": 1583368549977,
  "prev_events": [
    "$EVkBBcVBiPZqbPtrvQMegKtTZnNEAZc9gX03Gv8Q6hY",
    "$G9yuxNI3XWxr8Q6hWbzia52edKQsjzPUv4XPE_gtCiY",
    "$GMngiP_mNuWlXw7KbnKr945_lkbBKRidXVpuhNjkAYg",
    "$GYL4zxoh8vM4nSMKYojQQLcNcuhZ2xi7lftiWEd8bbM",
    "$GcHRr4clXktwDlyy2q6trZr2VQ7Q5afx_hAhjSzpa40",
    "$Gkbbm6vP4cUf6YzHhlGOexvDi9u8oJXuB7bZGBJVL6Y",
    "$Gkkit3UjsQXblKnMNNe7GQ6OkBgYGmXmBfMO8MwI0mU",
    "$GqZHARE3tUnsN9svzBlh9764qAPNmSNeHzf1HpPN3Qs",
    "$GwokmZC3CufQQDUntMIgc7WrjFOMYYkuBObCu3d8Iy4",
    "$GyZ_au_apwMvEQ2Obkub1pKfl4te5ajuy6-SGv24UY0",
    "$HhAZ3ckzntEfMxpokN7iCREib5KwB-Xvq6bhqrqWBDI",
    "$HnOhRkuJw1s4TkiVqZY74jn2R2txs0eNPV13nw2EbEI",
    "$IduvLk52HlV86qqKQa8a1CEmkNO9J2IKqs-8xgXodm8",
    "$Iq3NvYpr-axhHHN6HNkjBlStd9qUPPb0Xf5wYt42_Cg",
    "$LAdcfI4dWnlplWuAXYOGWymuOIIPZOfTeuknVGi-yl8",
    "$ObFntf_wmFPH8Qrg2N5H96maPh-EbnUH83MniQZuxlk"
  ],
  "room_id": "!TdAwENXmXuMrCrFEFX:maunium.net",
  "sender": "@jason:zemos.net",
  "state_key": "@matthew:matrix.org",
  "type": "m.room.member",
  "signatures": {
    "zemos.net": {
      "ed25519:zASw3QBm": "gwljqbt+TsRajRnhGRnJ/27giqLswbGMRe0AzOI/29HzBUdX5t94uO4gm6umhB5uRx3Iab1S2XETXgZWYl1BBw"
    }
  },
  "unsigned": {
    
  }
}
```

The spec states for invites:
>  Once the event has been signed by both the inviting homeserver and the invited homeserver, it can be sent to all of the servers in the room by the inviting homeserver.

In this case, zemos.net has signed the invite but matrix.org has not, therefore the event fails signature checks.

**Comment by neilalexander** [comment]
The panic is fixed by 8589f83, but joining that room still won't work because of a faulty state event. Dendrite seems to not drop the event correctly so will address that separately.

**Comment by neilalexander** [comment]
This will be fixed by #1407.
